### DISCUSSION TITLE
Compilation problem

### DISCUSSION THREAD

**Discussion by Alexus-3000** [josephernest/SamplerBox#68] [Q&A]
On several different systems, including Raspberry:

**sudo pip3 install git+https://github.com/SamplerBox/rtmidi-python.git**

Collecting git+https://github.com/SamplerBox/rtmidi-python.git
  Cloning https://github.com/SamplerBox/rtmidi-python.git to /tmp/pip-req-build-xfbk592i
  Running command git clone --filter=blob:none --quiet https://github.com/SamplerBox/rtmidi-python.git /tmp/pip-req-build-xfbk592i
  Resolved https://github.com/SamplerBox/rtmidi-python.git to commit f7b95708eb6a9fd2277518930aab41a43287fd91
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: rtmidi-python
  Building wheel for rtmidi-python (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [17 lines of output]
      running bdist_wheel
      running build
      running build_ext
      Compiling rtmidi_python.pyx because it changed.
      [1/1] Cythonizing rtmidi_python.pyx
      /usr/local/lib/python3.10/dist-packages/Cython/Compiler/Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-xfbk592i/rtmidi_python.pyx
        tree = Parsing.p_module(s, pxd, full_module_name)
      building 'rtmidi_python' extension
      creating build
      creating build/temp.linux-x86_64-3.10
      creating build/temp.linux-x86_64-3.10/RtMidi
      x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -D__LINUX_ALSA__ -I/usr/include/python3.10 -c RtMidi/RtMidi.cpp -o build/temp.linux-x86_64-3.10/RtMidi/RtMidi.o
      RtMidi/RtMidi.cpp:1048:10: fatal error: alsa/asoundlib.h: Нет такого файла или каталога
       1048 | #include <alsa/asoundlib.h>
            |          ^~~~~~~~~~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for rtmidi-python
  Running setup.py clean for rtmidi-python
Failed to build rtmidi-python
Installing collected packages: rtmidi-python
  Running setup.py install for rtmidi-python ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for rtmidi-python did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      running install
      /usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_ext
      building 'rtmidi_python' extension
      creating build
      creating build/temp.linux-x86_64-3.10
      creating build/temp.linux-x86_64-3.10/RtMidi
      x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -D__LINUX_ALSA__ -I/usr/include/python3.10 -c RtMidi/RtMidi.cpp -o build/temp.linux-x86_64-3.10/RtMidi/RtMidi.o
      RtMidi/RtMidi.cpp:1048:10: fatal error: alsa/asoundlib.h: Нет такого файла или каталога
       1048 | #include <alsa/asoundlib.h>
            |          ^~~~~~~~~~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> rtmidi-python

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.

**Comment by dosas** [comment]
If 'I' translate this line `RtMidi/RtMidi.cpp:1048:10: fatal error: alsa/asoundlib.h: Нет такого файла или каталога` correctly it says:

```
RtMidi/RtMidi.cpp:1048:10: fatal error: alsa/asoundlib.h: No such file or directory
```

This means that the dependecy `libasound2-dev` is missing. Installing it should fix your problem?

There were two PRs [[1]](https://github.com/josephernest/SamplerBox/pull/54), [[2]](https://github.com/josephernest/SamplerBox/pull/62) in the past trying to add this dependency amongst other changes. I have no idea why it became a dependency now.
